components/brave_wallet/browser/polkadot/polkadot_keyring.cc

```diff
@@ -103,4 +113,138 @@ std::optional<std::string> PolkadotKeyring::AddNewHDAccount(uint32_t index) {
   return GetAddress(index, IsTestnet() ? kWestendPrefix : kPolkadotPrefix);
 }
 
+void PolkadotKeyring::SetRandBytesForTesting(  // IN-TEST
+    const std::vector<uint8_t>& seed_bytes,
+    const std::vector<uint8_t>& nonce_bytes) {
+  CHECK_IS_TEST();
+  rand_seed_bytes_for_testing_ = seed_bytes;
+  rand_nonce_bytes_for_testing_ = nonce_bytes;
+}
+
+// Creates JSON to export Polkadot account info in the proper format.
+// At this time password is reused to encrypt encoded data.
+std::optional<std::string> PolkadotKeyring::EncodePrivateKeyForExport(
+    uint32_t account_index,
+    const std::string& password) {
+  if (password.empty()) {
+    return std::nullopt;
+  }
+
+  auto& keypair = EnsureKeyPair(account_index);
+  auto seed = keypair.GetExportKeyPkcs8();
+
+  // Get the address for this account.
+  std::string address = GetAddress(account_index, kSubstratePrefix);
+
+  // Generate a random salt for scrypt.
+  std::vector<uint8_t> salt(32);
+  if (rand_seed_bytes_for_testing_) {
+    CHECK_IS_TEST();
+    salt = rand_seed_bytes_for_testing_.value();
+  } else {
+    crypto::RandBytes(salt);
+  }
+
+  // Derive encryption key from password using scrypt
+  // Substrate/Polkadot standard parameters: n=32768, r=8, p=1.
+  std::array<uint8_t, crypto_secretbox_KEYBYTES> derived_key;
+  crypto::kdf::ScryptParams scrypt_params = {
+      .cost = 32768,                         // n
+      .block_size = 8,                       // r
+      .parallelization = 1,                  // p
+      .max_memory_bytes = 64 * 1024 * 1024,  // 64 MB
+  };
+
+  if (!crypto::kdf::DeriveKeyScryptNoCheck(
+          scrypt_params, base::as_byte_span(password),
+          base::span<const uint8_t>(salt), derived_key)) {
+    return std::nullopt;
+  }
+
+  // Generate a random nonce for xsalsa20-poly1305.
+  std::vector<uint8_t> nonce(crypto_secretbox_NONCEBYTES);
+  if (rand_nonce_bytes_for_testing_) {
+    CHECK_IS_TEST();
+    nonce = rand_nonce_bytes_for_testing_.value();
+  } else {
+    crypto::RandBytes(nonce);
+  }
+
+  // Encrypt the seed using NaCl secretbox (xsalsa20-poly1305).
+  std::vector<uint8_t> plaintext(seed.begin(), seed.end());
+  std::vector<uint8_t> padded_plaintext(crypto_secretbox_ZEROBYTES +
+                                        plaintext.size());
+  std::vector<uint8_t> ciphertext(padded_plaintext.size());
+
+  // Pad plaintext with zeros (NaCl secretbox requirement).
+  for (size_t i = 0; i < plaintext.size(); ++i) {
```

Comment:
Are you sure you need to conduct the padding here? Typically crypto_secretbox will handle the padding for you under the hood.

Reply:
I think yes, [brave_sync](https://github.com/brave/brave-core/blob/61e9aa127d68a91364b15172c00738f17feb3ae8/components/brave_sync/crypto/crypto.cc#L96) following same approach